File: backend/src/main/java/com/gca/service/GroupService.java
package com.gca.service;

import com.gca.dto.GroupDTO;
import com.gca.dto.TreeNodeDTO;

import java.util.List;

/**
 * Interfaz que define las operaciones del servicio de grupos.
 */
public interface GroupService {

    /**
     * Obtiene el árbol de grupos.
     *
     * @return un objeto TreeNodeDTO que representa la estructura del árbol de grupos
     */
    TreeNodeDTO getTree();

    /**
     * Obtiene todos los grupos.
     *
     * @return una lista de todos los TreeNodeDTO
     */
    List<TreeNodeDTO> getAllGroups();

    /**
     * Elimina un grupo por su ID.
     * @param id el ID del grupo a eliminar
     */
    void deleteGroup(Long id);

    /**
     * Crea un nuevo grupo.
     *
     * @param group los datos del grupo a crear
     * @return el ID del grupo creado
     */
    Long createGroup(GroupDTO group);

    /**
     * Actualiza un grupo existente.
     *
     * @param group los datos del grupo a actualizar
     * @return el ID del grupo actualizado
     */
    Long updateGroup(GroupDTO group);

    /**
     * Asigna una plantilla a un grupo.
     * @param groupId el ID del grupo al que se asignará la plantilla
     * @param templateId el ID de la plantilla a asignar
     */
    void assignTemplateToGroup(Long groupId, Long templateId);

    /**
     * Desasigna una plantilla de un grupo.
     * @param groupId el ID del grupo del que se desasignará la plantilla
     */
    void unassignTemplateToGroup(Long groupId);
}

File: frontend/src/app/services/group/group.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { TreeNodeDTO } from '../../models/tree-node.dto';
import { GroupDTO } from '../../models/group.dto';

/**
 * Servicio para la gestión de grupos
 */
@Injectable({
  providedIn: 'root'
})
export class GroupService {
  private http = inject(HttpClient);
  private apiUrl = environment.apiUrl;
  private readonly baseUrl = this.apiUrl + '/groups';

  /**
   * 
   * @param group Grupo a añadir
   * @returns 
   */
  addGroup(group: GroupDTO) : Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
    });
    return this.http.post<void>(this.apiUrl + "/registerGroup", group, { headers })
  }

  /**
   * 
   * @param group Grupo a modificar
   * @returns 
   */
  updateGroup(group: GroupDTO) : Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
    });
    return this.http.post<void>(this.apiUrl + "/updateGroup", group, { headers })
  }

  /**
   * 
   * @returns Devuelve todos los grupos
   */
  getAllGroups(): Observable<TreeNodeDTO[]> {
    return this.http.get<TreeNodeDTO[]>(this.baseUrl);
  }

  /**
   * 
   * @param id Identificador del grupo a eliminar
   * @returns 
   */
  deleteGroup(id:number): Observable<void> {
    return this.http.delete<void>(this.baseUrl + "/" + id)
  }

  /**
   * Asigna / Elimina una plantilla a un determinado grupo
   * @param groupId Identificador del grupo
   * @param templateId Identificador de la plantilla
   * @returns 
   */
  asignTemplateToGroup(groupId:number, templateId:number): Observable<void> {
    if (templateId) {
      return this.http.post<void>(this.baseUrl + "/" +groupId + "/assign-template/" + templateId, null)
    } else {
      return this.http.delete<void>(this.baseUrl + "/" + groupId + "/unassign-template" )
    }
  }
}